Dockerfile

```diff
@@ -1,46 +1,37 @@
-# Stage 1: Build stage - Used to prepare dependencies and package the application
-FROM python:3.11-slim-bookworm AS base
-
-# Install Java runtime (required for PySpark/Hail) and procps for process management
-RUN echo "deb http://deb.debian.org/debian oldstable main" >> /etc/apt/sources.list
-RUN apt-get update && \
-    apt-get install --no-install-recommends -y openjdk-11-jdk-headless procps && \
-    # Clean up apt cache to reduce image size
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/*
-
-FROM base AS builder
-
-# Add UV binaries
-COPY --from=ghcr.io/astral-sh/uv:python3.11-bookworm-slim /usr/local/bin/uv /usr/local/bin/
+FROM ghcr.io/astral-sh/uv:python3.12-bookworm-slim AS uv_builder
+ENV UV_COMPILE_BYTECODE=1 UV_LINK_MODE=copy
+# Disable python downloads to use the one from the base image
+ENV UV_PYTHON_DOWNLOADS=0
 
 # Set working directory for application files
 WORKDIR /app
 
-# Copy project configuration and source code
-ADD ./pyproject.toml  ./uv.lock /app/
-ADD ./src/ /app/src/
-# Create an empty README file (required by some package managers)
-RUN touch README.md
-
-# Install dependencies and project
 RUN --mount=type=cache,target=/root/.cache/uv \
-    uv sync --frozen --no-editable
-
-# Stage 2: Runtime stage - Creates the final minimal image
-FROM base AS production
+    --mount=type=bind,source=uv.lock,target=uv.lock \
+    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
+    uv sync --locked --no-install-project --no-dev
+COPY src /app/src
+COPY README.md /app/README.md
+COPY LICENSE.md /app/LICENSE.md
+COPY pyproject.toml /app/pyproject.toml
+COPY uv.lock /app/uv.lock
+RUN --mount=type=cache,target=/root/.cache/uv \
+    uv sync --locked --no-dev
 
-# Set working directory in the runtime container
-WORKDIR /app
+# # Stage 2: Runtime stage - Creates the final minimal image
+FROM python:3.12.11-slim-trixie AS production
 
-# Copy the virtual environment with all dependencies from the builder stage
-COPY --from=builder /app/.venv .venv
+# # Set working directory in the runtime container
+COPY --from=uv_builder --chown=app:app /app /app
```

Comment:
I don't get the concept of the different stages. How many images are we building here? It's not a cardinal issue, as this design was already in place.